flax/linen/linear.py

```diff
@@ -364,13 +372,35 @@ def __call__(self, inputs: Array) -> Array:
     kernel = self.param('kernel', self.kernel_init, kernel_shape)
     kernel = jnp.asarray(kernel, self.dtype)
 
+    if self.padding == 'CIRCULAR':
+      padding_lax = 'VALID'
+    else:
+      padding_lax = self.padding
+
     y = lax.conv_transpose(inputs,
                            kernel,
                            strides,
-                           self.padding,
+                           padding_lax,
                            rhs_dilation=self.kernel_dilation,
                            precision=self.precision)
 
+    if self.padding == "CIRCULAR":
```

Comment:
this is pretty complicated. Is there a reference for this?

Reply:
@jheek I didn't use any reference (except for verifying the results against `neural_tangents.stax.ConvTranspose`), but let me explain the logic here:
After `ConvTranspose` has been calculated with "VALID" padding into variable `y` in [this line](https://github.com/google/flax/pull/1661/files#diff-cfcdd6089c08daca13a27f79e07b856585c5f9b6910570f9cc3e7be76fd746b1L367), we need to
1. identify the size of the final output along each spatial dimension (let's call it "period")
2.  pad each dimension to a certain integer number of periods
3. wrap the array periodically around each spatial dimension

Step by step:

1. Size of the final output is stride * input size 

2. For padding, we need to understand how much to pad from left/right for each spatial dimension. Padding should satisfy three criteria:
- total size after padding should be an integer number of periods
- padding should be the symmetric (same from the right and from the left)
- the element corresponding to beginning of the original input data inside the padded array should be located at integer number of periods (otherwise we'll get correct answer, but circularly shifted)

The above is satisfied if we compute the padding like I did in the code:
- Compute the difference between  the size of `y` and the desired size of the final output
- Compute complement of this difference to even number of periods
- Divide the complement into two for left and right padding

3. After the padding is done, we can wrap the array around each spatial dimension. I did this separately for each dimension, reshaping the array into (..., -1, period, ...) and summing over the corresponding axis. Let me know if there is a simpler way to do it - I haven't found one (see e.g. https://stackoverflow.com/questions/42297115/numpy-split-cube-into-cubes) 

There is a subtlety with even-sized kernels: the choice to write `(size_diff + 1) // 2, size_diff // 2)` (and not `(size_diff // 2, (size_diff + 1) // 2)`) [here](https://github.com/google/flax/pull/1661/files#diff-cfcdd6089c08daca13a27f79e07b856585c5f9b6910570f9cc3e7be76fd746b1R397) reflects the alignment convention I talk about in [a comment above](https://github.com/google/flax/pull/1661#issuecomment-961904066)

What's the best way to proceed, should I add some of this description to inline comments?

Reply:
Yes I think it makes sense to add (a summary of) in inline comments

Reply:
@jheek Sure, I've expanded the comments in `ConvTranspose` - please have a look

Reply:
Thanks looks great!